tests/hashing/main.c

```diff
@@ -0,0 +1,373 @@
+        dividend *= 10;
+    }
+    buf[len] = '\0';
+
+    /* End here if rounding is not desired */
+    if (!round) {
+        return;
+    }
+
+    /* Do the rounding */
+    quot = dividend / divisor;
+    if (quot >= 5) {
+        for (int j=len-1; j>=0; j--) {
+            if (buf[j] == '.' || buf[j] == '-') {
+                continue;
+            }
+            if (buf[j] >= '9') {
+                buf[j] = '0';
+            } else {
+                buf[j]++;
+                break;
+            }
+        }
+        /* Special case if a number that was rounded up only contained the digit 9 */
+        if ((!is_negative && buf[0] == '0' && buf[1] == '0')
+            || (is_negative && buf[1] == '0' && buf[2] == '0')) {
+                for (size_t k=len-1; k>0; k--) {
+                    buf[k] = buf[k-1];
+                }
+                if (!is_negative) {
+                    buf[0] = '1';
+                } else {
+                    buf[1] = '1';
+                }
+                buf[len] = '\0';
+                if (buf[len-1] == '.') {
+                    buf[len-1] = '\0';
+                }
+        }
+    }
+}
+
+#ifdef SHOW_PROGRESS
+
+/*
+    This thread prints the current benchmarking progress
+    (i.e. the number of iterations finished)
+    Native board only
+*/
+void *prog_thread(void *arg)
+{
+    uint32_t *i = arg;
+    msg_t msg;
+
+    while (1) {
+        xtimer_sleep(1);
+        msg_try_receive(&msg);
+        if (msg.content.value == 1) {
+            msg.content.value = 0;
+            printf("\n");
+            thread_sleep();
+        } else if (msg.content.value == 2) {
+            printf("\n");
+            return NULL;
+        } else {
+            printf("Progress: %"PRIu32"\r", *i);
+        }
+    }
+}
+
+#endif
+
+int main(void) {
+
+    printf( "\n\nMeasure performance of SHA256 and Keccak in ticks needed to calculate "
+            "%d hash operations and in hash operations per tick.\n\n", num_iterations);
+
+    /* Initialize */
+    int32_t it_counter = 0;
+    xtimer_ticks64_t start_ticks;
+    xtimer_ticks64_t end_ticks;
+    uint64_t ticks_dif;
+    char ticks_buf[32];
+
+    /*
+        This thread prints the current benchmarking progress
+        (i.e. the number of iterations finished)
+        Native board only
+    */
+    kernel_pid_t prog_thread_pid =
+                thread_create(prog_thread_stack, sizeof(prog_thread_stack),
+                                THREAD_PRIORITY_MAIN - 1, THREAD_CREATE_SLEEPING,
+                                prog_thread, &it_counter, "prog_thread");
+    msg_t msg;
+
+    /* Iterate through all desired string lengths */
+    for (uint32_t k=0; k<sizeof(databytelens)/sizeof(databytelens[0]); k++) {
+
+        size_t databytelen = databytelens[k];
+
+        /* Generate random string */
+        printf("\nGenerating random string of length %d.\n\n", databytelen);
+        random_init(0x33799f);
+        char* datastring = malloc(databytelen);
```

Comment:
don't use malloc

Reply:
I had to use it here because I wouldn't have enough memory otherwise for the longer strings and get hard faults. What do you suggest instead?

Reply:
just allocate max memory needed upfront, I guess?

Reply:
There is not enough memory to statically allocate 102400 bytes on the native board (I get a segmentation fault) and on actual boards this would fail for less (e.g. 10240 bytes is already too much for the stm32f0discovery), which means I would need a way to find out what the maximum memory that I can allocate is. How could I do this?

This said, is this still relevant since you want me to remove the tests/hashing directory and its contents for this PR?